File: Queue/slidingWindowMax.java
class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
        int[] res = new int[nums.length - k + 1];
        LinkedList<Integer> l = new LinkedList<Integer>();

        for (int i = 0; i < nums.length; i++) {

            if (!l.isEmpty() && l.peek() < i - k + 1) { // not part of curr window
                l.poll();
            }

            while (!l.isEmpty() && nums[i] >= nums[l.peekLast()]) {
                l.pollLast(); // removeRear
            }

            l.offer(i);

            if (i - k + 1 >= 0)
                res[i - k + 1] = nums[l.peek()];
        }

        return res;
    }
}

File: SlidingWindow/slidingwindowmax.py
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        res = []
        q = collections.deque()

        for i in range(len(nums)):
            val = nums[i]

            if (q and q[0] < i - k + 1):
                q.popleft()
            
            while (q and val >= nums[q[-1]]):
                q.pop()

            q.append(i)

            print(q)

            if (i - k + 1 >= 0):
                res.append(nums[q[0]])
        return res
